### PR TITLE
SMT: Model Extraction & Pretty Counterexamples

### LINKED ISSUE #63: Pretty Print Counterexamples
The counterexamples are currently not pretty printed in any way, and look very messy. For exmaple for [this spec](https://github.com/ethereum/act/blob/b78ffd6715323dd93ee36aaf9d6035963e3e80f0/tests/invariants/fail/non-inductive.act):

```

Invariant "(C.x < 9)" of "C":



---



---



---

Counter example found!

SMTModel {modelObjectives = [], modelBindings = Nothing, modelAssocs = [("C_j_x",7 :: Integer),("C_j_x_post",100 :: Integer),("C_j_CALLER",0 :: Integer),("C_j_CALLVALUE",0 :: Integer),("C_j_CALLDEPTH",0 :: Integer),("C_j_ORIGIN",0 :: Integer),("C_j_BLOCKHASH","" :: String),("C_j_BLOCKNUMBER",0 :: Integer),("C_j_DIFFICULTY",0 :: Integer),("C_j_CHAINID",0 :: Integer),("C_j_GASLIMIT",0 :: Integer),("C_j_COINBASE",0 :: Integer),("C_j_TIMESTAMP",0 :: Integer),("C_j_THIS",0 :: Integer),("C_j_NONCE",0 :: Integer)], modelUIFuns = []}
```

We should tidy this up, perhaps something like the following:

```
Invariant C.x < 9 of C:

  Counterexample:

    prestate:

      calldata:

        j()
      
      storage:

        x = 7

      environment:

        THIS      = 0
        NONCE     = 0
        ORIGIN    = 0
        CHAINID   = 0
        GASLIMIT  = 0
        COINBASE  = 0
        TIMESTAMP = 0
        CALLVALUE = 0
        CALLDEPTH = 0

    poststate:

      x = 100
```

### PR DESCRIPTION
**Pull request by xwvvvvwx**
This closes #63.

Notable changes:

- Pretty printed counterexamples
- Reworked solver interaction code so that we now share a single long running solver instance between all queries. This is a little more involved, but makes extracting the models much easier
- The `Inv` constructor for the `Query` type has been reworked a little to bundle all the individual smt queries related to each invariant in one type
- The various `locsFrom*` / `ethEnvFrom*` functions in `Extract.hs` now deduplicate their output, making them more ergononomic at the cost of some redundent calls to `nub`

Pretty printing / colorful output is achieved using the [`ansi-wl-pprint`](https://hackage.haskell.org/package/ansi-wl-pprint-0.6.7.3/docs/Text-PrettyPrint-ANSI-Leijen.html#v:encloseSep) library. It may be worth reworking the pretty printer in `Print.hs` as well as the coq and k backends to make use of this, but I've left that for another day.

As I worked through this I realised that error handling in the SMT backend is also a little unsatisfactory, but since this is already quite a lot of code to review I've decided to tidy that up properly in another PR.

An example of a pretty printed counterexample:

```
    counterexample:

      calldata:

        f(x : 1, y : 0, z : true)

      storage:

        prestate:

          C.a : 0
          C.c : 0
          C.b : true

        poststate:

          C.a : 1
          C.c : 1
          C.b : false

      constructor arguments:

        constructor(x : 0, y : false, z : "")
```

### CODE REVIEW

File: src/test/Test.hs [hunk 1]
```diff
@@ -78,12 +78,18 @@ typeCheckSMT :: Solver -> GenT (Reader Bool) Property
...
-      parseOutput (exitCode, stdout, _) = case exitCode of
-        ExitFailure _ -> False
-        ExitSuccess -> not . any (isPrefixOf "(error") . filter (/= "") . lines $ stdout
+      runQueries smtconf queries = do
+        solverInstance <- spawnSolver smtconf
+        res <- mapM (askSMT solverInstance) queries
+        pure $ null (catMaybes res)
+
+      getSMT (Postcondition _ _ smt) = smt
+      getSMT _ = undefined
```
Comment (kjekac) [inline src/test/Test.hs:90]: Even though the input will always have been run through `mkPostconditionQueries` I would say that using `undefined` is bad practice in production code. (And `mkPostconditionQueries` may not do what the name suggests.) Maybe at least use `error "some nice msg"` instead?

File: src/SMT.hs [hunk 2]
```diff
@@ -96,7 +156,7 @@ data Model = Model
 --    - Asserts that the postcondition cannot be reached
 --   If this query is unsatisfiable, then there exists no case where the postcondition can be violated.
 mkPostconditionQueries :: Claim -> [Query]
-mkPostconditionQueries (B behv@(Behaviour _ _ _ (Interface ifaceName decls) preconds postconds stateUpdates _)) = mkQuery <$> postconds
+mkPostconditionQueries (B behv@(Behaviour _ Pass _ (Interface ifaceName decls) preconds postconds stateUpdates _)) = mkQuery <$> postconds
```
Comment (kjekac) [inline src/SMT.hs:159]: Why do we care about `Pass`?

File: src/Main.hs [hunk 3]
```diff
@@ -105,25 +107,63 @@ main = do
         contents <- readFile file'
         proceed contents (compile contents) $ \claims -> do
           let
-            handleRes (query, res) = case res of
-                  Unsat -> (True, typ <> " " <> prettyExp target <> " holds :)", show $ getSMT query)
-                  Sat -> (False, typ <> " " <> prettyExp target <> " does not hold :(", show $ getSMT query)
-                  SMT.Unknown -> (False, typ <> " " <> prettyExp target <> " could not be proved due to a solver timeout :(", show $ getSMT query)
-                  SMT.Error _ str -> (False, typ <> " " <> prettyExp target <> " could not be proved to due a solver error: " <> str, show $ getSMT query)
+            catModels results = [m | Sat m <- results]
```
Comment (kjekac) [inline src/Main.hs:110]: Maybe this is just me but I find this amount of nested `do`s, `let`s and `where`s to be quite difficult to follow. Many of these functions seem to be directly related to printing SMT stuff, and so maybe could be defined in `SMT.hs` instead?

File: src/Extract.hs [hunk 4]
```diff
@@ -3,18 +3,40 @@
...
+  Right (BytesUpdate item e) -> (BytesLoc item) : locsFromExp e
+
 locsFromReturnExp :: ReturnExp -> [StorageLocation]
 locsFromReturnExp (ExpInt e) = locsFromExp e
 locsFromReturnExp (ExpBool e) = locsFromExp e
 locsFromReturnExp (ExpBytes e) = locsFromExp e
 
 locsFromExp :: Exp a -> [StorageLocation]
-locsFromExp e = case e of
+locsFromExp e = nub $ case e of
```
[the remark this answers is not in the document]
Comment (xwvvvvwx) [reply src/Extract.hs:39]: nice, but `nub` is actually `O(n^2)` :(

We can get `O(n log(n))` if we have an `ord` instance for `Exp`.

File: src/SMT.hs (hunk 2 above)
Comment (xwvvvvwx) [reply src/SMT.hs:159]: This was a bug, postconditions should not hold for `Fail` queries (and actually thinking about it now, probably we shouldn't even include them when we split the act into pass / fail specs?).

File: src/Print.hs [hunk 5]
```diff
@@ -1,6 +1,6 @@
 {-# Language GADTs #-}
 
-module Print (prettyEnv, prettyExp, prettyType, prettyBehaviour, prettyItem) where
+module Print where
```
[the remark this answers is not in the document]
Comment (xwvvvvwx) [reply src/Print.hs:3]: well in the end I just thought that we wanted everything in the module to be exported and I was annoyed by that I kept having to add functions to the export list here.

Is there some agreed upon best practice for module import / export lists? We have quite a mix in our codebase at the moment...

File: src/Main.hs (hunk 3 above)
Comment (xwvvvvwx) [reply src/Main.hs:110]: Tidied up a bit here, think it should be a little clearer now.

**Comment by xwvvvvwx** [comment]
addressed review comments and rebased on master

File: src/Print.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (MrChico) [reply src/Print.hs:3]: I always found the explicit export function and import function kinda verbose and beaurocratic

File: src/SMT.hs [hunk 6]
```diff
@@ -59,36 +82,75 @@ data SMTExp = SMTExp
...
+      calldata' = (text "calldata:") <$$> line <> (indent 2 $ formatSig ifaceName args)
+      environment' = (text "environment:") <$$> line <> (indent 2 . vsep $ fmap formatEnvironment environment)
+      storage = (text "storage:") <$$> (indent 2 . vsep $ [ifExists prestate (line <> prestate'), poststate'])
+      initargs' = (text "constructor arguments:") <$$> line <> (indent 2 $ formatSig "constructor" initargs)
+
+      prestate' = (text "prestate:") <$$> line <> (indent 2 . vsep $ fmap formatStorage prestate) <> line
+      poststate' = (text "poststate:") <$$> line <> (indent 2 . vsep $ fmap formatStorage poststate)
+
+      formatSig iface cd = (text iface) <> (encloseSep lparen rparen (text ", ") $ fmap formatCalldata cd)
+      formatCalldata ((Decl _ name), val) = text $ name <> " : " <> prettyReturnExp val
```
Comment (MrChico) [inline src/SMT.hs:142]: I'd maybe go for `=` instead of `:` here. I associate `:` with typing

File: src/Main.hs [hunk 7]
```diff
@@ -105,25 +106,49 @@ main = do
...
-
-          pcResults <- mapM (runQuery config) (concatMap mkPostconditionQueries claims)
-          invResults <- mapM (runQuery config) (mkInvariantQueries claims)
-          let results = map handleRes (pcResults <> invResults)
-          allGood <- foldM (\acc (r, msg, smt) -> do
-            if (_debug config) then putStrLn (msg <> "\n\n" <> smt) else putStrLn msg
-            pure $ if acc == False then False else r
-            ) True results
-          unless allGood exitFailure
+                holds = and (fmap isPass results)
```
Comment (kjekac) [inline src/Main.hs:128]:
```suggestion
                holds = all isPass results
```

File: src/Main.hs [hunk 8]
```diff
@@ -105,25 +106,49 @@ main = do
...
+                msg = identifier query <+> if holds then passMsg else failMsg results
+                smt = if debug' then line <> (getSMT query) else empty
```
Comment (kjekac) [inline src/Main.hs:130]:
```suggestion
                smt = if debug' then line <> getSMT query else empty
```

**Review by kjekac — COMMENTED** [review]
I unfortunately still haven't gone through `SMT.hs` properly but here are a few other things I found.

File: src/Main.hs (hunk 3 above)
Comment (kjekac) [reply src/Main.hs:109]: Yes, much better!

File: src/Print.hs (hunk 5 above)
[the remark this answers is not in the document]
Comment (kjekac) [reply src/Print.hs:3]: I've no idea lol. I think it makes sense to restrict exports when you're building EDSLs or datatypes which should have very specific interfaces but on our case I think it's fine to just export everything most of the time.

File: src/SMT.hs (hunk 6 above)
Comment (kjekac) [reply src/SMT.hs:142]: I'd change it to `=` for environment and storage too, for the same reason.

File: src/SMT.hs [hunk 9]
```diff
@@ -59,36 +82,75 @@ data SMTExp = SMTExp
...
+    where
+      calldata' = (text "calldata:") <$$> line <> (indent 2 $ formatSig ifaceName args)
+      environment' = (text "environment:") <$$> line <> (indent 2 . vsep $ fmap formatEnvironment environment)
+      storage = (text "storage:") <$$> (indent 2 . vsep $ [ifExists prestate (line <> prestate'), poststate'])
+      initargs' = (text "constructor arguments:") <$$> line <> (indent 2 $ formatSig "constructor" initargs)
+
+      prestate' = (text "prestate:") <$$> line <> (indent 2 . vsep $ fmap formatStorage prestate) <> line
+      poststate' = (text "poststate:") <$$> line <> (indent 2 . vsep $ fmap formatStorage poststate)
+
+      formatSig iface cd = (text iface) <> (encloseSep lparen rparen (text ", ") $ fmap formatCalldata cd)
```
Comment (kjekac) [inline src/SMT.hs:141]:
```suggestion
      calldata' = text "calldata:" <$$> line <> (indent 2 $ formatSig ifaceName args)
      environment' = text "environment:" <$$> line <> (indent 2 . vsep $ fmap formatEnvironment environment)
      storage = text "storage:" <$$> (indent 2 . vsep $ [ifExists prestate (line <> prestate'), poststate'])
      initargs' = text "constructor arguments:" <$$> line <> (indent 2 $ formatSig "constructor" initargs)

      prestate' = text "prestate:" <$$> line <> (indent 2 . vsep $ fmap formatStorage prestate) <> line
      poststate' = text "poststate:" <$$> line <> (indent 2 . vsep $ fmap formatStorage poststate)

      formatSig iface cd = text iface <> (encloseSep lparen rparen (text ", ") $ fmap formatCalldata cd)
```

**Review by kjekac — COMMENTED** [review]
This looks great! :) Some of my suggestions are definitely nitpicks and with some exceptions, I think most aren't about correctness or even best practices, but more about style. But I figured you'd appreciate to at least have seen those even if you don't want to apply them?

Other than that, I'd suggest that we start adding comments/documentation to any functions we touch in a PR? I believe I've done so in my last two PRs and intend to continue. Would you mind adding some here too? 😬

File: src/Extract.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (kjekac) [reply src/Extract.hs]: Damn, of course. Let's deal with that later, maybe? One single `O(n^2)` seems fine for now.

File: src/SMT.hs (hunk 6 above)
Comment (kjekac) [reply src/SMT.hs:142]: Oh and you also don't need the parentheses here 🙃: 

```suggestion
      formatCalldata (Decl _ name, val) = text $ name <> " : " <> prettyReturnExp val
```

File: src/SMT.hs [hunk 10]
```diff
@@ -224,42 +286,226 @@ mkInvariantQueries claims = concatMap mkQueries gathered
...
+  _ <- sendCommand solverInstance "(set-option :print-success true)"
+  err <- sendLines solverInstance smtPreamble
+  case err of
+    Nothing -> pure solverInstance
+    Just msg -> error $ "could not spawn solver: " <> msg
+
+stopSolver :: SolverInstance -> IO ()
+stopSolver (SolverInstance _ stdin stdout stderr process) = cleanupProcess (Just stdin, Just stdout, Just stderr, process)
+
+sendLines :: SolverInstance -> [SMT2] -> IO (Maybe String)
```
[the remark this answers is not in the document]
Comment (xwvvvvwx) [reply src/SMT.hs:349]: yeah, this is definately weird, and should definately be fixed by the error handling rework.

**Review by kjekac — APPROVED** [review]
I'm into this. :) Comments would still be good (guessing you're working on those atm?) but the code itself seems good to go imo!

**Comment by kjekac** [comment]
This looks great to my mind :)

**Pull request merged**